File: college-clubs-mobile/lib/models/event.dart
class Event {
  final int id;
  final int clubId;
  final String title;
  final String? description;
  final String venue;
  final String formattedDate;
  final double price;
  final String status;
  final int capacity;
  final int spotsLeft;
  final String? imagePath;
  final String? volunteers;
  final List<String> allImages;

  Event({
    required this.id,
    required this.clubId,
    required this.title,
    this.description,
    required this.venue,
    required this.formattedDate,
    required this.price,
    required this.status,
    required this.capacity,
    required this.spotsLeft,
    this.imagePath,
    this.volunteers,
    required this.allImages,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    // Safely extract first image path if it exists
    String? firstImage;
    List<String> imagesList = [];
    
    if (json['images'] != null && (json['images'] as List).isNotEmpty) {
      firstImage = json['images'][0]['path'];
      imagesList = (json['images'] as List)
          .map((img) => (img['path'] ?? '').toString())
          .where((path) => path.isNotEmpty)
          .toList();
    }

    return Event(
      id: json['id'],
      clubId: json['club_id'],
      title: json['title'] ?? '',
      description: json['description'],
      venue: json['venue'] ?? json['place'] ?? 'Main Campus',
      formattedDate: json['formatted_date'] ?? json['date_string'] ?? 'Date Pending',
      price: double.parse((json['price'] ?? 0.00).toString()),
      status: json['status'] ?? 'completed',
      capacity: json['capacity'] ?? 100,
      spotsLeft: json['spots_remaining'] ?? json['capacity'] ?? 100,
      imagePath: firstImage,
      volunteers: json['volunteers']?.toString(),
      allImages: imagesList,
    );
  }
}

File: college-clubs/app/Models/Event.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Event extends Model
{
    use HasFactory;

    protected $fillable = [
        'club_id',
        'title',
        'description',
        'venue',
        'place',
        'start_time',
        'end_time',
        'date_string',
        'volunteers',
        'price',
        'status',
        'capacity',
    ];

    protected $casts = [
        'start_time' => 'datetime',
        'end_time' => 'datetime',
        'price' => 'decimal:2',
        'capacity' => 'integer',
    ];

    public function club(): BelongsTo
    {
        return $this->belongsTo(Club::class);
    }

    public function images(): HasMany
    {
        return $this->hasMany(EventImage::class);
    }

    public function registrations(): HasMany
    {
        return $this->hasMany(Registration::class);
    }

    // Helper to check if event is full
    public function isFull(): bool
    {
        return $this->registrations()->where('status', '!=', 'cancelled')->count() >= $this->capacity;
    }

    // Helper to get spots remaining
    public function spotsRemaining(): int
    {
        $activeRegistrationsCount = $this->registrations()->where('status', '!=', 'cancelled')->count();
        return max(0, $this->capacity - $activeRegistrationsCount);
    }

    // Helper to print a unified event date representation
    public function getFormattedDateAttribute(): string
    {
        if ($this->start_time) {
            return $this->start_time->format('M d, Y @ h:i A');
        }

        return $this->date_string ?? 'Date Pending';
    }

    // Helper to print a unified event venue/place representation
    public function getFormattedVenueAttribute(): string
    {
        return $this->venue ?? ($this->place ?? 'Venue Pending');
    }
}
